app/javascript/elm/src/Main.elm

```diff
@@ -249,15 +241,50 @@ update msg model =
                 )
 
         DeselectSession ->
-            ( { model | selectedSessionId = Nothing }, Ports.checkedSession { deselected = model.selectedSessionId, selected = Nothing } )
+            case model.selectedSession of
+                Success selectedSession ->
+                    ( { model | selectedSession = NotAsked }, Ports.checkedSession { deselected = Just selectedSession.id, selected = Nothing } )
+
+                _ ->
+                    ( model, Cmd.none )
 
         ToggleSessionSelectionFromAngular maybeId ->
             case maybeId of
                 Just id ->
-                    ( { model | selectedSessionId = Just id }, Cmd.none )
+                    ( model, SelectedSession.fetch model.selectedSensorId model.page id GotSession )
 
                 Nothing ->
-                    ( { model | selectedSessionId = Nothing }, Cmd.none )
+                    ( { model | selectedSession = NotAsked }, Cmd.none )
+
+        ToggleSessionSelection id ->
+            case model.selectedSession of
+                NotAsked ->
+                    ( model
+                    , Cmd.batch
+                        [ Ports.checkedSession { deselected = Nothing, selected = Just id }
+                        , SelectedSession.fetch model.selectedSensorId model.page id GotSession
+                        ]
+                    )
+
+                Success selectedSession ->
+                    if selectedSession.id == id then
+                        ( { model | selectedSession = NotAsked }
+                        , Ports.checkedSession { deselected = Just selectedSession.id, selected = Nothing }
+                        )
+
+                    else
+                        ( { model | selectedSession = NotAsked }
+                        , Cmd.batch
+                            [ Ports.checkedSession { deselected = Just selectedSession.id, selected = Just id }
```

Comment:
I'm wondering whether another name wouldn't be better. "checked" is ambiguous, I'd consider using "selected" which is consistent with the other naming (if it makes sense) or sth like "chosen" or "toggled". I feel like for sure I've had this conversation before but don't remember the outcome :P but it may be different this time anyway

Reply:
Good one! Renamed to `toggleSession`. At some point we may want to have a chat about naming ports. Not sure if it's better the imperative or past tense form (ie toggleSession vs sessionToggled)